server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -295,6 +306,57 @@ public void onFailure(Exception e) {
         });
     }
 
+    static final class PrepareTargetResult {
+        final boolean isSequenceNumberRecovery;
+        final long startingSeqNo;
+        final TimeValue openEngineTime;
+        final Store.MetadataSnapshot recoveryTargetMetadata;
+
+        PrepareTargetResult(boolean isSequenceNumberRecovery, long startingSeqNo,
+                            TimeValue openEngineTime, Store.MetadataSnapshot recoveryTargetMetadata) {
+            this.isSequenceNumberRecovery = isSequenceNumberRecovery;
+            this.startingSeqNo = startingSeqNo;
+            this.openEngineTime = openEngineTime;
+            this.recoveryTargetMetadata = recoveryTargetMetadata;
+        }
+    }
+
+    /**
+     * Prepares the recovery target for sequence number based recovery if possible. If the recovery source has all required operations
+     * up to the {@link StartRecoveryRequest#startingSeqNo()}, then it can perform sequence number based recovery without coordinating
+     * with the recovery target. However, if its history is only up to the global checkpoint on the target, it needs to ask the target
+     * to locally recover up to the global checkpoint then perform a sequence number based recovery. If the target can recover properly,
+     * then it will leave its engine open for a sequence number recovery; otherwise it needs to close its engine, capture and send back
+     * the latest metadata to the recovery source for a file-based recovery.
+     */
+    void prepareTargetForSequenceBasedRecovery(ActionListener<PrepareTargetResult> listener) throws IOException {
+        if (request.startingSeqNo() != SequenceNumbers.UNASSIGNED_SEQ_NO && isTargetSameHistory()) {
+            if (shard.hasCompleteHistoryOperations("peer-recovery", request.startingSeqNo())) {
+                listener.onResponse(new PrepareTargetResult(true, request.startingSeqNo(), null, request.metadataSnapshot()));
+            } else {
+                final long startingSeqNo = request.globalCheckpoint() + 1;
+                if (request.startingSeqNo() < startingSeqNo && shard.hasCompleteHistoryOperations("peer-recovery", startingSeqNo)) {
+                    startEngineOnTarget(
```

Comment:
If we set GCP = LCP of the received safe commit, then that commit is no longer safe if it has holes in sequence numbers. Therefore, we won't perform a sequence-number based recovery if the previous recovery did not complete. However, I think this is a good suggestion as it's okay to perform a file-based recovery in this case. I would like to hear from others before heading towards this direction. @ywelsch @DaveCTurner WDYT?

Reply:
I don't really follow precisely what kinds of "previous recovery did not complete" we're discussing, and wonder if both instances of that phrase above are talking about the same kinds of failure. Can you clarify? I don't think we should be too heroic here in any case. This may also change with @ywelsch's [suggestion](https://github.com/elastic/elasticsearch/pull/43463/files#r296233244) (with which I agree) to always try a local recovery first.

Reply:
If the main worry is about the 2 numbers in start recovery request, we can perhaps just use the global checkpoint as the starting seq no that we send with start recovery request (i.e. we are optimistic that we can locally recover up to that point, even if we later find out that we don't have all the ops in the translog). When the primary then does phase0 (tells replica to recover locally from translog up to gcp), and it notices that not all docs are there, it sends a revised starting seq no as part of the phase0 response, which the primary can then take into account.

Reply:
A healthy replica should be able to replay until its own global check point. If it cannot, would it not be just as good to assume it is corrupted and fallback to file based recovery?

Reply:
corrupted is perhaps too strong of a word. Right now, a replica ends up in this situation after phase 1. It has a safe commit, but does not have the corresponding translog yet to replay up to the global checkpoint. I think it's fine to fall back to file based recovery in that case. What we need to be careful about then is that if we recover from a primary that predates this PR, that we're not sending the global checkpoint as starting seq no, but the local checkpoint of safe commit, as before.

Reply:
I am thinking about a slightly different option. Before sending the start recovery request, the recovery target opens an engine, recovers locally up to the global checkpoint, flushes, and closes the engine. After these preliminary steps, the recovery target will perform the exact flow we have today (i.e., use the local checkpoint of the safe commit as the starting seq_no). The downside of this approach is that the local recovery is useless if a file-based occurs.

Reply:
I have updated this PR to recover locally before sending the start recovery request. It's pretty contained. Please have a look and let me know your thought. Thank you.